opencog/atoms/execution/EvaluationLink.cc

```diff
@@ -740,7 +740,7 @@ TruthValuePtr do_eval_with_args(AtomSpace* as,
 
 		// Be sure to specify the atomspace in which to work!
 		PythonEval &applier = PythonEval::instance();
-		return applier.apply_tv(as, schema.substr(pos), args);
+		return applier.apply_tv(schema.substr(pos), args);
```

Comment:
This is surely a bug ... during pattern matching, possible matches are evaluated in a temporary atomspace. If, in the end, the match is good, then the results are placed in the main atomspace. If there is no match, the results are discarded.  With the above change ... where are the temporary results written? How can you know which atomspace to use?

Reply:
I'm pretty sure that none of the python unit tests test for this, since the pattern matcher is almost never used with python.  But if you go through the unit tests and look for `GroundedPredicate "scm:foo"`  (which calls foo in the temp atomspace) and then create a matching `GroundedPredicate "py:foo"`, you'll see the failure.

Reply:
Which test should it port to python to expose bug?

Reply:
Not sure, but here's how to find out: in `DefaultPatternMatchCB` you will find 
```
   _temp_aspace = grab_transient_atomspace(as);
   _instor = new Instantiator(_temp_aspace);
```
and then, below, in `DefaultPatternMatchCB::eval_term` you will see 
```
gvirt = HandleCast(_instor->instantiate(virt, gnds, true));
```
which uses the temp atomspace, and also two diferent uses of
```
EvaluationLink::do_eval_scratch(_as, gvirt, _temp_aspace, true);
```
put an assert in there, and see which unit tests trigger this.  All users of `EvaluationLink::do_eval_scratch()` will be using temp atomspaces.

Reply:
Or just put the assert into `EvaluationLink::do_eval_scratch()` and see which unit tests trigger that.

Reply:
So for example ... `do_eval_with_args` is *always* called with the temp atomspace, from what I can tell, so the `apply_tv()` will always specify the temp atomspace.

Reply:
I ported CondLinkUTest::test_grounded_cond. It passed, i still do no see why should it fail. Pattern matcher just checks TruthValue returned.. TruthValue might not be in any atomspace at all.

Reply:
From the code which Anatoly changed the only usage of this parameter was to set it to `_atomspace` field of the `PythonEval`. And only usage of `_atomspace` field was to set it as global `ATOMSPACE` variable in Python modules. And global `ATOMSPACE` variable is not used by Python code. So may be intention was as you described but in fact this code doesn't work by this way.